### ISSUE TITLE
Make maxBadConnRetries configurable

### ISSUE THREAD

**Issue by thameezb** [golang/go#48309]
### What version of Go are you using (`go version`)?
v1.16.2

### Does this issue reproduce with the latest release?
Yes

### What operating system and processor architecture are you using (`go env`)?
n/a 

### What did you do?
The value of `maxBadConnRetries` is hardcoded to 2  -> https://github.com/golang/go/blob/master/src/database/sql/sql.go#L1511. 
I was wondering if there is a specific reason for this, as there seems to be attempts to make this variable in the past which were abandoned -> https://github.com/golang/go/issues/15608

Making it variable would be very much appreciated in our particular use-case. 

#### Scenario
- Running a k8s cluster with various services running as pods in the cluster. Each of these pods has a sidecar Health-Monitor container which simply pings certain dependencies on a regular interval (10s). Should the health-check fail, the dependency is marked as down and potentially the pod can be marked as not-ready. 
- One of our dependency types is SQL, and the go-sql lib is used. 
- There are times where we need to bring down our DB (during patching etc) 
  - and during these times we see connections to the DB spike to triple in traffic (makes sense now that we have looked at the code)
  -  this causes the DB instance to be hit hard and load to increase, during a period where load is already high due to an external process 
  - this also means that the HM acts differently according to different scenarios (happy/sad)
 
Note that we cannot simply block connections to the entire DB instance as it needs to be contactable during this period.

**Comment by seankhliao** [comment]
cc @bradfitz @kardianos

**Comment by thameezb** [comment]
any update?

**Comment by kardianos** [comment]
There are two ways to approach this, especially that we do a better job of marking connections as bad before they go into the pool to begin with.

 1. Add a configuration with the driver to not return driver.ErrBadConn
 2. Add a configuration with database/sql to not retry.

I'm open to (2), but I would prefer it not to introduce more locks and shared state. Want to suggest a change?

**Comment by kardianos** [comment]
@sjmudd `database/sql` already uses `isBadConn = errors.Is(err, driver.ErrBadConn)` [1]. It is encouraged for database drivers to wrap BadConn.

[1] https://cs.opensource.google/go/go/+/refs/tags/go1.19.5:src/database/sql/sql.go;l=852

**Comment by sjmudd** [comment]
Yes, and I would like it to be incorporated into https://github.com/go-sql-driver/mysql but believe there are some concerns about compatibility issues with the different versions of go that is supported.

I would like also to see exposed something like:
- a counter of non-logical errors, perhaps like [DBStats](https://pkg.go.dev/database/sql#DBStats) which I guess you could use, and maybe counted by the type of error, so the user can identify both the fact an error has occurred and also which types they were.
- the errors could be handled by the low level drivers but that does not make much sense to me. It feels better that they can return them to database/sql and your package makes them visible in a standard way to the user.
- e.g. something like `DBErrors` being a `map[errorType] -> struct { Count int, LastError: error, RetriedCount int }` or similar.

I believe that there was some concern about making `maxBadConnRetries` configurable due to potential locking issues. Is that still the case? If so you may not be keen to have to maintain and additional DBErrors structure.  However, under normal circumstances the actual number of errors I'd expect to see of type `driver.ErrBadConn` would be rather small so I would expect locking issues to not be a major concern.

What I see now is about 1 in 100,000 connections failing though that's after the retry having been applied so I would expect the actual error count to be higher. Given I'm polling a large number of servers all the time this number is actually significant as some of these connections are actually to healthy servers which need to be fixed when another server fails. My primary interest is to be able to collect statistics from database/sql and optionally be able to turn off the retries. With that in place I can then get the detail necessary back from a modified [The MySQL driver](https://github.com/go-sql-driver/mysql) to identify what specific conditions are generating these errors.

**Comment by sjmudd** [comment]
Slightly related but of relevance is that MySQL had [a patch](https://bugs.mysql.com/bug.php?id=93240) provided by a colleague for this specific problem which now makes "updated clients" see an error message in such a case (e.g. adjust your MySQL setting of `wait_timeout` as you've been idle for too long...).  I believe the go MySQL driver layer does not handle this (and I need to make bug report for that, official MySQL clients/libraries from Oracle do but they don't provide a go driver) and as things stand even if the error was reported up to the database/sql layer in the way it's done now it would trigger a reconnect to the database server and the error would never be seen.

So having a way to configure the retries, and having visibility on the problem when all things are in place would allow people monitoring errors more closely to see errors and then act upon them and thus have a better experience.
